doc/devguide/extending/app-layer/app-layer-frames.rst

```diff
@@ -0,0 +1,227 @@
+
+Using the *SMB* parser as example, before frame support, a rule would look like::
+
+    alert tcp ... flow:to_server; content:"|ff|SMB"; content:"some smb 1 issue";
+
+With frame support, one is able to do::
+
+    alert smb ... flow:to_server; frame:smb1.data; content:"some smb 1 issue";
+
+Implementation & API Callbacks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Rust
+----
+
+The frame types are defined as an enum. In Rust, derive from the ``AppLayerFrameType``:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: #[derive(AppLayerFrameType)]
+    :lines: 1-15
+
+How some frames are registered in the `SMB <https://github.com/OISF/suricata/blob/master/rust/src/smb/smb.rs#L1383>`_ parser:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: pub fn parse_tcp_data_ts<'b>
+    :end-at: nbss_hdr.data, nbss_hdr.length as i64);
+    :lines: 1-3, 56-74
+    :dedent: 4
+
+These are the frame registration functions highlighted above:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: fn add_nbss_ts_frames(&mut
+    :end-before: fn add_smb1_ts_hdr_data
+    :dedent: 4
+
+Frame registering with the `telnet <https://github.com/OISF/suricata/blob/master/rust/src/telnet/telnet.rs#L171>`_ parser:
+
+.. _telnet-snippet:
+
+.. literalinclude:: ../../../../rust/src/telnet/telnet.rs
+    :caption: rust/src/telnet/telnet.rs
+    :language: rust
+    :start-at: fn parse_request
+    :end-at: TelnetFrameType::Data as u8)
+    :lines: 1, 19-29
+    :dedent: 4
+
+The parameters represent:
+
+- ``flow``: dedicated data type, carries specific flow-related data
+- ``stream_slice``: dedicated data type, carries stream data, shown further bellow
+- ``frame_start``: a pointer to the start of the frame buffer in the stream (``cur_i`` in the SMB code snippet)
+- ``frame_len``: what we expect the frame length to be (the engine may need to wait until it has enough data. See what is done in the telnet snippet request frames registering)
+- ``frame_type``: type of frame it's being registering (defined in an enum, as shown further above)
+
+.. note:: on frame_len
+        
+    For protocols which search for an end of frame char, like telnet, indicate unknown length by passing ``-1``. Once the length is known, it must be updated. For those where length is a field in the record (e.g. *SMB*), the frame is set to match said length, even if that is bigger than the current input
+
+``StreamSlice`` contains the input data to the parser, alongside other Stream-related data important in parsing context. Definition  is found in *applayer.rs*:
+
+.. literalinclude:: ../../../../rust/src/applayer.rs
+    :caption: rust/src/applayer.rs
+    :language: rust
+    :start-at: pub struct StreamSlice
+    :end-before: impl StreamSlice
+
+Registering relevant frame callbacks:
```

Comment:
can we add that these are inferred from the frame enum's `AppLayerFrameType` derive statement?